consensus/cometbft/service/finalize_block.go

```diff
@@ -94,9 +98,66 @@ func (s *Service) finalizeBlockInternal(
 		TxResults:             txResults,
 		ValidatorUpdates:      valUpdates,
 		ConsensusParamUpdates: &cp,
+		NextBlockDelay:        s.nextBlockDelay(req),
 	}, nil
 }
 
+// Panics if SBT upgrade parameters are not set.
+//
+//nolint:lll // long message on one line for readability.
+func (s *Service) nextBlockDelay(req *cmtabci.FinalizeBlockRequest) time.Duration {
+	delay := constBlockDelay
+
+	switch {
+	case s.blockDelay == nil: // height > 0; blockDelay record doesn't exist in DB (first live upgrade)
+		if s.sbtUpgradeHeight == 0 {
+			panic("SBT (stable block time) upgrade detected. --beacon-kit.cometbft.sbt-upgrade-height (block's height where SBT was enabled) must be set")
+		}
+		switch {
+		case req.Height > s.sbtUpgradeHeight: // upgrade happened in the past
+			if s.sbtUpgradeTime.IsZero() {
+				panic("SBT (stable block time) upgrade detected. --beacon-kit.cometbft.sbt-upgrade-time (block's time where SBT was enabled) must be set")
+			}
+			s.blockDelay = BlockDelayUponGenesis(
+				s.sbtUpgradeTime,
+				s.sbtUpgradeHeight,
+			)
+			delay = s.blockDelay.Next(req.Time, req.Height)
+		case req.Height == s.sbtUpgradeHeight: // upgrade is happening now
+			s.blockDelay = BlockDelayUponGenesis(
+				req.Time,
+				req.Height,
+			)
+			// default: // upgrade is in the future
+			//   // nothing to do
+		}
+	case s.blockDelay.InitialHeight < s.sbtUpgradeHeight: // height > 0; blockDelay record exists in DB; consecutive (2nd, 3rd, so on) live upgrade
+		// Calculate block delay based on existing checkpoint.
+		delay = s.blockDelay.Next(req.Time, req.Height)
+
+		switch {
+		case req.Height > s.sbtUpgradeHeight: // upgrade happened in the past
+			if s.sbtUpgradeTime.IsZero() {
+				panic("SBT (stable block time) upgrade detected. --beacon-kit.cometbft.sbt-upgrade-time (block's time where SBT was enabled) must be set")
+			}
+			s.blockDelay = BlockDelayUponGenesis(
+				s.sbtUpgradeTime,
+				s.sbtUpgradeHeight,
+			)
+			delay = s.blockDelay.Next(req.Time, req.Height)
+		case req.Height == s.sbtUpgradeHeight: // upgrade is happening now
+			s.blockDelay = BlockDelayUponGenesis(
```

Comment:
unlike the case `s.blockDelay == nil` here we would return `delay = s.blockDelay.Next(req.Time, req.Height)` instead of constBlockDelay right? 

Wouldn't this result in different delays being used?

Reply:
> unlike the case s.blockDelay == nil here we would return delay = s.blockDelay.Next(req.Time, req.Height) instead of constBlockDelay right?

Right. 

Before the first upgrade, we always use `constBlockDelay = 500ms` (before `s.sbtUpgradeHeight + 1` is reached). Then it's whatever you set `TargetBlockTime` to (e.g. 2s). 

If you change the target block time again, the subsequent block delay will be calculated using the previous checkpoint (up until `s.sbtUpgradeHeight + 1`).

Reply:
Thinking about this more, we should probably save `TargetBlockTime` to the db as well.